Add watermark option

Emilien-Foissotte opened the pull request:
Hi @SiddhantSadangi  ! 👋 

Here is the fixed PR, sorry for the misguidance on the old PR (#5 )

I've needed recently in the need of adding watermark on some PDF documents to mask PII or avoid reuse of documents, e.g. on ID Cards, pay slips, etc (in the fun process of contacting owners and real estate agencies to find a new flat 🫠)

I've used pdf-workdesk to reduce PDF size and I crafted a small evolution on the app to apply a watermark on the whole page based on a text.
Here is a preview, I ran it locally to test the app on my CV 📖

![Screenshot 2025-03-29 at 17-12-14 PDF WorkDesk](https://github.com/user-attachments/assets/687e4715-79dd-4601-a67c-e3b7de82cc54)


I hope this suggestion will be useful for the community 🙌

Feel free to make openly your review and your suggestion on implementation !  

**PS : I'v locked again the requirements file using [uv](https://docs.astral.sh/uv/). But feel free to dismiss the `uv.lock` and `pyproject.toml` if it doesn't suit your projects needs / development process.**

## Summary by Sourcery

Add a watermarking feature to PDF WorkDesk, allowing users to add custom text watermarks to PDF documents

New Features:
- Implement a PDF watermarking functionality that allows users to add custom text watermarks diagonally across PDF pages

Enhancements:
- Update the application UI to include a new watermark option with text input and font size slider
- Modify the README to include watermarking in the list of supported features

Build:
- Add pyproject.toml for project configuration
- Update dependencies using uv package manager

Chores:
- Update requirements to use the latest package versions

## Summary by Sourcery

Add a watermarking feature for PDFs with UI controls and merge it into the app, lock dependencies, update UI toggles, bump version, and refresh documentation.

New Features:
- Enable users to add custom text watermarks across PDF pages with configurable size, color, and transparency

Enhancements:
- Replace several PDF reduction checkboxes with toggles for consistency
- Increase sidebar HTML height to display additional sponsor badge
- Move PDF watermark controls into its own expander with text input, sliders, and download button

Build:
- Lock and pin all project dependencies in requirements.txt using uv

Documentation:
- Update README to list Watermark feature, adjust feature headers, and add GitHub Sponsor badge
- Add GitHub Sponsor badge to sidebar HTML

Chores:
- Bump project version from 0.6.2 to 0.7.0

SiddhantSadangi commented:
Hey @Emilien-Foissotte 👋 

Thank you so much for this PR! 🤗 

I've been wanting to add this feature to the app for quite a while now but have been keeping busy with other personal commitments (moving to a new house being one of them 😅 )

I'll need a while to take a deeper look at the changes, but meanwhile, could you please address the issues raised by the auto-reviewer? I went through them, and they sound relevant for an app built to be used in a concurrent environment.

Hopefully, once the changes are merged, we won't need to use https://filigrane.beta.gouv.fr/ anymore 😆

SiddhantSadangi commented on utils/helpers.py:
```diff
@@ -459,3 +462,45 @@ def convert_pdf_to_word(pdf):
+    packet = BytesIO()
+    can = canvas.Canvas(packet, pagesize=letter)
+    can.setFont("Helvetica", stamp_size)
+    can.setFillColor(red)
```
The default `red` looks too bright. Can we let users select the color using `st.color_picker` instead, and if possible, add some transparency to it?

Emilien-Foissotte replied:
Yes of course, I will add options to let the user decide the color and transparency 👍

Emilien-Foissotte replied:
Solved in the last version 👍

Emilien-Foissotte commented:
> Hey @Emilien-Foissotte 👋 
> 
> Thank you so much for this PR! 🤗 
> 
> I've been wanting to add this feature to the app for quite a while now but have been keeping busy with other personal commitments (moving to a new house being one of them 😅 )
> 
> I'll need a while to take a deeper look at the changes, but meanwhile, could you please address the issues raised by the auto-reviewer? I went through them, and they sound relevant for an app built to be used in a concurrent environment.
> 
> Hopefully, once the changes are merged, we won't need to use https://filigrane.beta.gouv.fr/ anymore 😆 

Thanks for the feedback 🙏 Very glad to see that it's useful and that it was expected in your app roadmap! 🙌

Good luck with your new house btw 🤞 I will consider the feedbacks from auto-reviewer, they quite accurate and relevant 👍

Hopefully yes ! 😆

Emilien-Foissotte commented on app.py:
```diff
@@ -247,6 +247,25 @@
+            text_watermark = st.text_input("Enter watermark text")
+            size_watermark = st.slider("Font size", min_value=6, max_value=30, value=12)
+
+            helpers.watermark_pdf(
```
Solved in the last version 👍

Emilien-Foissotte commented on utils/helpers.py:
```diff
@@ -459,3 +462,45 @@ def convert_pdf_to_word(pdf):
+        writer.add_page(page)
+
+    # Write the output PDF
+    with open("watermarked.pdf", "wb") as f:
```
Solved in the last version 👍

Emilien-Foissotte commented on utils/helpers.py:
```diff
@@ -459,3 +462,45 @@ def convert_pdf_to_word(pdf):
+    can.save()
+    # Save the watermark PDF
+    packet.seek(0)
+    with open("watermark_canva.pdf", "wb") as f:
```
Solved in the last version 👍

Emilien-Foissotte commented on utils/helpers.py:
```diff
@@ -459,3 +462,45 @@ def convert_pdf_to_word(pdf):
     return docx_stream
+
+
+@st.cache_data
```
Cache function should work as expected, no underlying file modifications expected

Emilien-Foissotte commented:
Hi again @SiddhantSadangi, 

I fixed all relevant concerns raised by Sourcery and recommended by you also ! (I didn't know this sourcery, pretty convenient and accurate !)

In this last version, it's possible for the user to pick a color and add transparency to the watermark. 

Regarding the concurrency warning, indeed I modified the implementation to use buffers instead of dumped files on the local filesystem 👍 

![Screenshot 2025-03-30 at 14-32-00 PDF WorkDesk](https://github.com/user-attachments/assets/9137ec77-508b-4d25-ab21-adc4b7b14d8f)

I hope it suits your vision of the feature, let me know your feedback openly ! 🙏

SiddhantSadangi commented:
@sourcery-ai review

SiddhantSadangi commented:
Hey @Emilien-Foissotte !

Thanks. The changes look a lot better now!

I ran a couple more cycles of the auto-reviewer and committed some of the changes suggested.

Some of the open conversations look helpful. Could you take a look, please?

P.S: I love [Sourcery](https://sourcery.ai/), it is a boon for OSS maintainers!

Emilien-Foissotte commented on utils/helpers.py:
```diff
@@ -459,3 +461,54 @@ def convert_pdf_to_word(pdf):
     return docx_stream
+
+
+@st.cache_data
```
Done in the last version

Emilien-Foissotte commented on app.py:
```diff
@@ -247,35 +247,31 @@
                     )
+        # create a watermark
+        with lcol.expander("©️ Add watermark"):
+            text_watermark = st.text_input("Enter watermark text")
```
Implemented in the last version

Emilien-Foissotte commented on utils/helpers.py:
```diff
@@ -459,3 +461,54 @@ def convert_pdf_to_word(pdf):
+
+
+@st.cache_data
+def watermark_pdf(
```
Implemented in the last version

Emilien-Foissotte commented on utils/helpers.py:
```diff
@@ -459,3 +461,54 @@ def convert_pdf_to_word(pdf):
+    # convert color from tuple to RGB color
+    color = tuple(int(stamp_color[i : i + 2], 16) / 256 for i in (0, 2, 4))
+    # apply transparency
+    color = (*color, stamp_transparency)
```
Implemented in the last version

Emilien-Foissotte commented:
Hi @SiddhantSadangi , 

I've implemented the suggestion made by Sourcery 👍 

Here are a few remarks about the overall comments made by Sourcery : 

> Consider adding a configuration option to control the watermark angle.

&

> Consider adding error handling to the watermark function to gracefully handle invalid color codes or other unexpected inputs.

To me, It seems overengineering to ask the user about an angle and validating color codes, as color picker handle only valid codes. Keeping it simple this way seems better in my opinion !  

And regarding 
> The dependency updates look extensive; consider whether a smaller, more targeted update would suffice.

Regarding the requirement update, it's safer to pin all requirements (uv / poetry way of freezing all deps) instead of pinning only the top ones. The `pyproject.toml` dependencies are the top ones on which the app depends logically on.

SiddhantSadangi commented:
@Emilien-Foissotte - sorry for the delayed response.. been keeping a bit busy.
I agree with your comments, no need to make this more complex ✅ 

I'll try to do a manual review this weekend 📝

Emilien-Foissotte commented:
> @Emilien-Foissotte - sorry for the delayed response.. been keeping a bit busy.
> I agree with your comments, no need to make this more complex ✅ 
> 
> I'll try to do a manual review this weekend 📝 

Hi @SiddhantSadangi !

Do you have some interesting feedback to add to sourcery PR, in case you might have some time with French bank holidays ? 🌞

Have a good 1st May !

SiddhantSadangi commented:
Hey @Emilien-Foissotte,
I am so sorry I haven't had the chance to look into this yet. I am currently prioritising another major project and hope to get it done by the end of May. I'll get back to this PR once that is done.

Is this PR blocking you by any chance?

Emilien-Foissotte commented:
Absolutely not, I was just nudging in case there was a chance to review the PR next days !

Good luck with your other project, we'll see this later, no problem! 🙌

SiddhantSadangi commented:
@sourcery-ai review

Emilien-Foissotte commented:
Hi @SiddhantSadangi ! 👋

Thanks for the update and review, everything seems good to me 😊

SiddhantSadangi commented:
Merged 🚀 

Again, thank you so much for your contribution and patience! 🙌
